### ISSUE TITLE
Incorrect CPU resource enumeration in LXD containers 

### ISSUE THREAD

**Issue by ilhaan** [rancher/rancher#21057]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
Make sure cluster is set to use "Custom" providers upon cluster creation.  Create LXD container with Docker installed as per instructions [here](https://memcpy.io/running-docker-privileged-inside-of-lxc-lxd.html). `ubuntu:18.04` is the base image that was used. Once container is ready, run "Custom Node Run Command" to enlist the container as a node on Rancher. This command can be found by navigating to "Cluster" in the top navbar > On the cluster page click the vertical ellipsis button > Edit > Scroll all the way to the bottom of the page. Copy this command and run it in the LXD container that was created earlier. 

**Result:**
Nodes successfully enlist in cluster and can schedule pods without any issues. However, nodes display CPU resources of the bare metal the LXD containers run on and not the resources (CPU Cores & Memory) that have been allocated to the LXD container. 

**Other details that may be helpful:**
LXD containers were ensured to have limited CPU cores using `lxc config set my-container limits.cpu 4`. This was verified by running `grep -c processor /proc/cpuinfo`, which returned `4` as expected. 

Rancher shows the same node as having 56 cores, which is the number of cores on the bare metal where the LXD containers are running. This is shown in the screenshot below: 

![Screen Shot 2019-06-21 at 1 14 32 PM](https://user-images.githubusercontent.com/5787928/59948956-94f0fc00-9426-11e9-9b60-3a58844cdd66.png)

All nodes shown in the screenshot are individual LXD containers running on the same machine. They have each been assigned 4 cores and 8GB of memory using LXD config settings. 

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): 2.2.4 (from bottom left in the UI) 
- Installation option (single install/HA): Single Install of rancher server using instructions from [here](https://rancher.com/quick-start/). 

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Custom 
- Machine type (cloud/VM/metal) and specifications (CPU/memory): metal
- Kubernetes version (use `kubectl version`): 

```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.3", GitCommit:"5e53fd6bc17c0dec8434817e69b04a25d8ae0ff0", GitTreeState:"clean", BuildDate:"2019-06-07T09:55:27Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.5", GitCommit:"2166946f41b36dea2c4626f90a77706f426cdea2", GitTreeState:"clean", BuildDate:"2019-03-25T15:19:22Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`): Below is the output from Docker running in LXD container 

```
Client:
 Version:           18.09.6
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        481bc77
 Built:             Sat May  4 02:35:57 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.6
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       481bc77
  Built:            Sat May  4 01:59:36 2019
  OS/Arch:          linux/amd64
  Experimental:     false

```

**Comment by alena1108** [comment]
@ilhaan Rancher just displays the information from k8s node status (allocatable/capacity fields). By default, the allocated resource is set to the node’s capacity. To set it to a particular value, you have to reserve https://stupefied-goodall-e282f7.netlify.com/contributors/design-proposals/node/node-allocatable/#kube-reserved or/and https://stupefied-goodall-e282f7.netlify.com/contributors/design-proposals/node/node-allocatable/#system-reserved flags. They can be passed as extra_args to the kubelet via cluster edit as yaml option.

**Comment by ilhaan** [comment]
@alena1108 I just tried this and updated my cluster's yaml file to the following: 
```
  kubelet: 
    fail_swap_on: false
    kubeReserved:
        cpu: "4"
        memory: "8Gi"
```
After adding the `kubeReserved` section, I clicked "Save". However, when I check the yaml config again the change does not seem to have persisted. I also didn't receive any updates from the UI telling me that this change has been made. Anything I am doing wrong here?

**Comment by ilhaan** [comment]
So after spending some time looking into this, I finally have a solution!

I noticed that LXD containers themselves had the correct (container aware) resource values in `/proc` but Docker containers running inside LXD containers did not, as shown below: 

```
root@lxdcontainer:~# grep -c processor /proc/cpuinfo 
1
root@lxdcontainer:~# grep MemTotal /proc/meminfo 
MemTotal:         488280 kB
root@lxdcontainer:~# docker run -it ubuntu bash 
root@fcd00479a2c1:/# grep -c processor /proc/cpuinfo 
56
root@fcd00479a2c1:/# grep MemTotal /proc/meminfo 
MemTotal:       263758600 kB
```

I posted about this on the LXD forums [here](https://discuss.linuxcontainers.org/t/lxcfs-with-security-nesting-true/5730) and discovered that this occurred due to Docker not being aware of [LXCFS](https://linuxcontainers.org/lxcfs/introduction/) and mounting its own copy of `/proc`. 

LXCFS can be used with Docker containers as shown [here](https://github.com/lxc/lxcfs#using-with-docker). Also, according to [this](https://rancher.com/docs/rke/latest/en/config-options/services/services-extras/#extra-binds) extra binds can be added to services. So I made an edit to my cluster and added the following to `kubelet`: 

```
kubelet: 
    extra_binds: 
      - "/var/lib/lxcfs/proc/cpuinfo:/proc/cpuinfo:rw"
      - "/var/lib/lxcfs/proc/diskstats:/proc/diskstats:rw"
      - "/var/lib/lxcfs/proc/meminfo:/proc/meminfo:rw"
      - "/var/lib/lxcfs/proc/stat:/proc/stat:rw"
      - "/var/lib/lxcfs/proc/swaps:/proc/swaps:rw"
      - "/var/lib/lxcfs/proc/uptime:/proc/uptime:rw"
```

After doing this and letting my cluster upgrade, Rancher shows the correct resource values for CPU and Memory on all cluster nodes! 

Thank you @superseb for sharing the `extra_binds` document with me on the Slack channel.
